src/ripple/protocol/ApiVersion.h

```diff
@@ -0,0 +1,132 @@
+ * because all the RPC requests with a version >= 2 must
+ * explicitly specify the version in the requests.
+ * Note that apiVersionIfUnspecified will be lower than
+ * apiMinimumSupportedVersion when we stop supporting API
+ * version 1.
+ *
+ * Command line Requests use apiCommandLineVersion.
+ */
+
+namespace RPC {
+
+template <unsigned int Version>
+constexpr static std::integral_constant<unsigned, Version> apiVersion = {};
+
+constexpr static auto apiInvalidVersion = apiVersion<0>;
+constexpr static auto apiVersionIfUnspecified = apiVersion<1>;
+constexpr static auto apiMinimumSupportedVersion = apiVersion<1>;
+constexpr static auto apiMaximumSupportedVersion = apiVersion<2>;
+constexpr static auto apiCommandLineVersion =
+    apiVersion<1>;  // TODO Bump to 2 later
+constexpr static auto apiBetaVersion = apiVersion<3>;
+constexpr static auto apiMaximumValidVersion = apiBetaVersion;
+
+static_assert(apiMinimumSupportedVersion >= apiVersionIfUnspecified);
+static_assert(apiMaximumSupportedVersion >= apiMinimumSupportedVersion);
+static_assert(apiBetaVersion >= apiMaximumSupportedVersion);
+static_assert(apiMaximumValidVersion >= apiMaximumSupportedVersion);
+
+}  // namespace RPC
+
+template <unsigned minVer, unsigned maxVer, typename Fn, typename... Args>
+    requires(maxVer >= minVer) &&                   //
+    (minVer >= RPC::apiMinimumSupportedVersion) &&  //
+    (RPC::apiMaximumValidVersion >=
+     maxVer) void forApiVersions(Fn const& fn, Args&&... args) requires requires
+{
+    fn(std::integral_constant<unsigned int, minVer>{},
+       std::forward<Args>(args)...);
+    fn(std::integral_constant<unsigned int, maxVer>{},
+       std::forward<Args>(args)...);
+}
+{
+    constexpr auto size = maxVer + 1 - minVer;
+    [&]<std::size_t... offset>(std::index_sequence<offset...>)
+    {
+        (((void)fn(
+             std::integral_constant<unsigned int, minVer + offset>{},
+             std::forward<Args>(args)...)),
+         ...);
+    }
+    (std::make_index_sequence<size>{});
+}
+
+template <typename Fn, typename... Args>
+void
+forAllApiVersions(Fn const& fn, Args&&... args) requires requires()
+{
+    forApiVersions<
+        RPC::apiMinimumSupportedVersion,
+        RPC::apiMaximumValidVersion>(fn, std::forward<Args>(args)...);
+}
+{
+    forApiVersions<
+        RPC::apiMinimumSupportedVersion,
+        RPC::apiMaximumValidVersion>(fn, std::forward<Args>(args)...);
+}
+
+struct ApiVersionHelper final
+{
+    constexpr static std::size_t size =
+        RPC::apiMaximumValidVersion + 1 - RPC::apiMinimumSupportedVersion;
+
+    constexpr static auto
+    index(unsigned int v) noexcept -> std::size_t
+    {
+        return static_cast<std::size_t>(v <= 1 ? 0 : (v <= 2 ? 1 : 2));
+    }
+
+    constexpr static auto
+    valid(unsigned int v) noexcept -> bool
+    {
+        return v >= RPC::apiMinimumSupportedVersion &&
+            v <= RPC::apiMaximumValidVersion;
+    }
+};
```

Comment:
We can, but how do you propose we manage our internal dependencies (levelization) versions ? Do you suggest we should move the content of `MultivarJson.h` inside `ApiVersion.h` ? It certainly can be done.

Reply:
`ApiVersion.h` is in `protocol` and not in `rpc`, because pulling dependencies from `rpc` into `protocol` would violate levelization ; we need API versions for doing anything version specific inside `protocol` (as we do inside `NetworkOPs.cpp`, baring some serious refactoring).

Reply:
There was this dependency here https://github.com/XRPLF/rippled/pull/4833/files#diff-86ea7ba43fce0b564af71cb3f2070ce71cb27d9ec9b9fd93ab5298960a3b36dbL71 which I glad to have removed, would rather not bring anything like it back.